### PR TITLE
export functions for setting init values

### LINKED ISSUE #9: higher-level helpers for parameters
[ This is just a stub for something that we've mentioned a lot (most recently in gh-5) but don't have a dedicated issue for. ]

From the README:

> At this point, nmrec is focused on parsing records into objects that are very closely tied to the form in the control stream and that don't lose information. In the future (and as needed), the plan is to add more convenience wrappers for working with these objects and helpers for getting things in and out of higher-level forms.

One of the main things this is referring to is the omega, sigma, and theta values.  The first step we be really sitting down and mapping out what the design should look like.

### PR DESCRIPTION
**Pull request by kyleam**
This series implements the nmrec-side for the bbr "inherit inits" feature (metrumresearchgroup/bbr#582, metrumresearchgroup/bbr#608).  It adds three functions for setting parameter init values: `set_theta`, `set_omega`, and `set_sigma`.

### Series overview

```
[01/12]  theta parser: fix handling of no-init form with spaces
[02/12]  matrix parser: parse values(diag,odiag) to nested option
[03/12]  define matrix helpers
[04/12]  get_record_option: extract selection logic to helper
```

Patches 1-4 are prep work.  Functionality-wise, patch 3 is the most interesting.  It adds some general helper functions that will be used for working the OMEGA and SIGMA input and creating the map from matrix indices to the record options.

```
[05/12]  add helper for mapping from parameter space to option
```

This adds the main behind-the-scenes logic for "indexing" parameter records (gathering the required info and making the parameter space to option map).  The docstring of `create_param_index` is the best place to start to get a high-level overview of what's going on there.

```
[06/12]  add functions to set theta, omega, and sigma init values
[07/12]  set_param: mention prior records if aborting due to length mismatch
[08/12]  set_param: allow overriding default format specifier
[09/12]  set_{omega,sigma}: support resetting representation
[10/12]  set_theta: support discarding bounds
[11/12]  set_param: drop parens if option was reduced to single item
```

Patches 6-11 introduce the user-exported functions and make various extensions.  Patch 9 is probably the one that deserves the most attention (and is something we haven't discussed), but of course any comments on the interface overall would be very appreciated.

```
[12/12]  README: drop paragraph about higher-level forms
```

Housekeeping.

---

I'm marking this as a draft because I'd like to go over it with fresh eyes (especially the docs, which I haven't read through again) and will likely reroll it with minor changes.  However, please feel free to start to take a look and comment.  @barretk it's in a state where I think you can begin wiring it up to bbr whenever you're ready.

cc @seth127

Closes #9.

### CODE REVIEW

**Comment by barrettk** [comment]
@kyleam I think we may want different handling when keeping the bounds. If two values are found, I would think we'd want to add a middle value as opposed to replacing the higher value:

```r
> test_case$input_nmrec
$PROBLEM theta bounds - maintain bounds; add starting value if not present

$THETA
(0, 2)      ; KA
(0, 3)      ; CL
(0, 10)     ; V2
(0.02)      ; RUVp
(1)         ; RUVa
(0, 0.5, 1) ; F
(-10, -3.92E+00, 10)
> test_case$replacement
[1] 1 2 3 4 5 6 7

> nmrec::set_theta(test_case$input_nmrec, test_case$replacement)
> test_case$input_nmrec
$PROBLEM theta bounds - maintain bounds; add starting value if not present

$THETA
(0, 1)      ; KA
(0, 2)      ; CL
(0, 3)     ; V2
(4)      ; RUVp
(5)         ; RUVa
(0, 6, 1) ; F
(-10, 7, 10)
```
Referring to the above example, I think we'd want to return the second record:
image

**Comment by kyleam** [comment]
@barrettk:

> I think we may want different handling when keeping the bounds. If two values are found, I would think we'd want to add a middle value as opposed to replacing the higher value:

The two-value form with one comma is `(low, init)`.  Why would we want to move to insert a value in between (i.e. move `init` to `up`, adding another `init`) rather than replacing `init`?

> Additionally, it would be ideal if we could maintain parentheses and any other formatting when discarding the bounds (second record in image below again):

Why you think that would be ideal?

This behavior (dropping parens if we've reduced a multi-value options down to one value) comes with ab7fab1.  As mentioned in that message, this applies to cases where we are dropping things at the request of a non-default option, so I think it's reasonable to simplify the result.  

Given that either way we go in some cases the result may or may not be consistent with neighboring single-value items, I lean toward giving the simplified result.  However, if you and @seth127 prefer retaining the parens in these cases, I'm okay with ejecting ab7fab1 from this series.

**Comment by barrettk** [comment]
>The two-value form with one comma is (low, init). Why would we want to move to insert a value in between (i.e. move init to up, adding another init) rather than replacing init?

Ahh I think I misunderstood this specification, though it sounds familiar from years ago when I was first learning about NONMEM. I was thinking the bounds were (low, hi) if two values were provided, but I [confirmed](https://nmhelp.tingjieguo.com/$theta.htm) the two-form you mentioned.

>Why you think that would be ideal?

I would have said it would make the tests nicer/easier using the framework we made in `bbr`, but since we're moving away from that, my main motivation is just the sentiment of wanting to change as little as possible from the original control stream file, and ideally only update the values. It doesn't really feel like a simplification from a non-technical (perhaps scientist) perspective, though I get what you're saying in your commit message. I would ask for @seth127's opinion on this as well though.


>Even accepting that we can rely on prior records, that seems like non-trivial logic that we'd be better off not adding unless there was a strong reason why someone shouldn't just use the informative names. (However, as far as I understand, a PRIOR subroutine can be used instead of a PRIOR record, so looking at the PRIOR record doesn't cover everything.)

> If someone wants to use set_{theta,omega,sigma} (or more likely the bbr wrapper), requiring them to transition the control stream to the informative names seems like a small ask, and their model code will be more readable for it.

Im slightly conflicted here. In terms of where I saw the examples involving the older method of specifying priors, I saw a notable number of examples. There are also cases where your error would trigger, but priors are not used. Here is one such [example](https://github.com/metrumresearchgroup/bbr/blob/main/inst/model/nonmem/complex/example2_saemimp.ctl), and related [test](https://github.com/metrumresearchgroup/bbr/blob/ac4293a28e001ce0977d76ff75c1f003ab85ca34/inst/test-refs/inherit-estimates/theta-fix-estimate.R) we had. This is because the second `$THETA 4 FIX` block is actually used as an alternative method for specifying `$OMEGAPD`. Though I agree with the sentiment 'we should push scientists to be more explicit and be rigid with what NONMEM syntax we support', I think scientists (and especially clients that use `bbr`) would end up running into this concern somewhat frequently. Given that, my vote would be to support cases where the expected number of replacements doesnt match the found number of records (including priors and the above case I mentioned).

**Comment by kyleam** [comment]
> It doesn't really feel like a simplification from a non-technical (perhaps scientist) perspective

Simplification is referring to reducing `(X)` to `X` (think of simplifying a math expression), not making any larger claims about what's simpler conceptually.

For a user standpoint, if you want to say `(1,2,3)`, you have no choice but to add the parens, but now the user is requesting 1 and 2 be dropped, we can't say whether they would have written that as `(2)` or `2` (and you can find lots of examples one way or the other). 

> my main motivation is just the sentiment of wanting to change as little as possible from the original control stream file, and ideally only update the values

Given we're touching the value anyway, I think it makes sense to go with the simplified form (but I'm just repeating myself).  If others share your preference, we can drop ab7fab1.

> What do you think about passing `representation` to `set_theta` as well, and have that argument toggle a similar feature?

I'd rather remove the "drop parens" simplification entirely than bloat the interface for a cosmetic detail.

**Comment by kyleam** [comment]
> Im slightly conflicted here. In terms of where I saw the examples involving the older method of specifying priors, I saw a notable number of examples.

In real project code or NONMEM examples?

> Though I agree with the sentiment 'we should push scientists to be more explicit and be rigid with what NONMEM syntax we support', I think scientists (and especially clients that use `bbr`) would end up running into this concern somewhat frequently.

I'm not convinced that it's a problem to tell them to switch to the the informative names if they want to use the "inherit inits" feature.

**Comment by barrettk** [comment]
With regards to the parentheses simplification, I'll defer to Seth. I see the argument for and against it, and would honestly lean towards what Seth thinks scientists would prefer/expect (but dont think it's significant enough to require scientist feedback)

>In real project code or NONMEM examples?

Both. Im not sure I could back track to see where I got specific examples from, as I just went through the repos I had to find example cases I had seen before. I think could be something worth asking @callistosp or someone about to get a better feel for how frequent this is seen in:
 - client models (past 3-5 years)
 - older Metrum projects
 - newer projects
My assumption is that Metrum scientists use the newer, more explicit version of specifying priors now, but that we have a lot of older models that still use the older method (unsure if we need to care about supporting these/how far back this starts). Im also under the assumption that many clients fall under the same boat, and if anything, are further behind in that transition.

Im of the opinion that **if we determined/ballparked** _~10% of older models and 20% of client models use the older method_, it's worth supporting the older method. 

>I'm not convinced that it's a problem to tell them to switch to the the informative names if they want to use the "inherit inits" feature.

I think **_if_** scientists ran into it somewhat frequently, it would be better to support it. I know we develop these packages with mostly _our_ scientists in mind, but given that we've given `bbr` demos to outside scientists that are transitioning to `R`/learning to code, I like the idea of reducing the burden on the scientist when possible. The more we can worry about syntax and they can worry about the math/science, the better. That being said, it seems like it could be worth getting scientist feedback before moving further with any refactor

**Comment by callistosp** [comment]
My two cents:

- drop the parens if you are dropping the bounds. They aren't needed at that point and I think most users don't include parens around boundless initial estimates

- I think most scientists use the new $THETAP/$OMEGAP naming conventions when they are creating a **new** model. However, if they are using an older model which was developed before this syntax was introduced, they may still have that old syntax included in their model. The new syntax was introduced in NONMEM 7.3 which was released in 2015. I would assume the vast majority of our users will be using NONMEM 7.3 or newer so there shouldn't by anything preventing them from using the THETAP/OMEGAP syntax. 
TL;DR: I don't think it's an unreasonable ask for the user to update any prior blocks to use the THETAP/OMEGAP syntax if they want to use the update inits function.

**Comment by barrettk** [comment]
@callistosp Thanks so much for your feedback! Discussing with Seth later today and that will be helpful in coming to a decision.

File: tests/testthat/test-set-param.R [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+        "$omega BLOCK(3) VALUES(1.2, 1.3)"
+      ),
+      values = matrix(
+        c(
+          2, 0, 0, 0,
+          0, 3, 0, 0,
+          0, NA, 0, 0,
+          0, 0, 0, 0
+        ),
+        nrow = 4, byrow = TRUE,
```
Comment (barrettk) [inline tests/testthat/test-set-param.R:274]: Perhaps I missed a commit message or docs, but I was just wondering what the motivation was for adding support for skipping values via `NA` was. It's nice to have that functionality in `nmrec`, but did you envision a way to implement this in `bbr`?

Comment (kyleam) [reply tests/testthat/test-set-param.R:274]: It seemed like a natural and straightforwad way to support updating only a subset of values for callers that are using these functions directly.

I didn't envision bbr exposing this functionality.

File: R/set-param.R [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+
+  if (length(values) != len_expected) {
+    abort(
+      c(
+        sprintf(
+          "Expected length of `values` to be %d, got %d.",
+          len_expected, length(values)
+        ),
+        "i" = sprintf(
+          "If you're using %s records for priors, please switch to informative record names.",
```
[the remark this answers is not in the document]
Comment (kyleam) [reply R/set-param.R:196]: sure (will wait to any other comments to settle)

**Review by seth127 — APPROVED** [review]
This is all looking good to me at this point. I think we have agreement on both points discussed above:

* Don't support old method of specifying priors
* Drop parens when no longer needed

@kyleam @barrettk did either of you have any outstanding comments or anything on this? If not, I think we're good to merge.

**Comment by seth127** [comment]
In terms of next steps here, I think we probably want a release sooner than later here, right? My motivation is that we'll want to gate the `nmrec` version on the accompanying `bbr` function, so we can't really move forward with https://github.com/metrumresearchgroup/bbr/pull/623 until this is actually _released_. Do both of you agree that that's the right approach?

**Comment by kyleam** [comment]
> @kyleam @barrettk did either of you have any outstanding comments or anything on this?

No, from my standpoint this is ready to merge.  (I'll push a dev version bump on top before merging and tag.)

> In terms of next steps here, I think we probably want a release sooner than later here, right?

Yes.  I'll plan to do a release this week.  There are one or two small things I may try to squeeze in (not functional changes), but that won't hold things up.

> My motivation is that we'll want to gate the `nmrec` version on the accompanying `bbr` function

A version gate is fine with me.  Even better in my view would be gating on the presence of one of these functions, but I know we do the version gate plenty.

> so we can't really move forward with [metrumresearchgroup/bbr#623](https://github.com/metrumresearchgroup/bbr/pull/623) until this is actually _released_.

I'd say you can move forward in terms of getting that CI wired up to test against this version and merging the PR.  It's just that a bbr _release_ is held up by this actually being released.

**Comment by barrettk** [comment]
Just to echo what Seth mentioned - this all looks good to me. Feel free to merge whenever
